### PR TITLE
[script][performance] Reworked the whole non-worn instrument routine, and added an exit if…

### PR DESCRIPTION
**Pull request by MahtraDR**
… no instrument can be found.

And added mindstate_goal command line argument, so this supersedes/closes https://github.com/rpherbig/dr-scripts/pull/5177 as well.

### CODE REVIEW

File: performance.lic [hunk 1]
```diff
@@ -2,43 +2,54 @@
...
+	{ name: 'mindstate_goal', regex: /(\d+)/, optional: true, description: 'Train until mindstate >= mindstate_goal' }
       ]
     ]
 
     args = parse_args(arg_definitions)
 
     @settings = get_settings
-    @song_list = get_data('perform').perform_options
+    song_list = get_data('perform').perform_options
+    @mindstate_goal = args.mindstate_goal.to_i
```
Comment (rpherbig) [inline performance.lic:24]:
```suggestion
    @mindstate_goal = args.mindstate_goal.to_i || @settings.performance_mindstate_goal
```

Comment (rpherbig) [reply performance.lic:24]: And then line 107 can be more or less unchanged

File: performance.lic [hunk 2]
```diff
@@ -52,37 +63,64 @@ class Performance
...
       end
     end
   end
 
+  def remove_hand_armor_and_get_instrument
+    if DRCI.remove_item?(@hand_armor)
+      @hand_armor_removed = true
+      DRCI.stow_item?(@hand_armor)
+    end
+    if DRCI.get_item?(@instrument)
```
Comment (rpherbig) [inline performance.lic:91]: Won't this try and get an instrument even if it shouldn't be removed (e.g. zills)? I don't see any guards in here

File: performance.lic [hunk 3]
```diff
@@ -52,37 +63,64 @@ class Performance
...
         if done_training?
+          wear_hand_armor_and_put_away_instrument
           exit
         else
-          play_song?(@settings, @song_list, is_instrument_worn, skip_clean)
+          DRC.play_song?(@settings, song_list, is_instrument_worn, skip_clean)
         end
       when /Perhaps you should find somewhere drier/
-        exit
+        if done_training?
```
Comment (rpherbig) [inline performance.lic:78]: Why is this now checking if `done_training?` - it used to always exit. Is it even possible to keep training when this message is seen?

File: performance.lic [hunk 4]
```diff
@@ -2,43 +2,54 @@
...
     @settings = get_settings
-    @song_list = get_data('perform').perform_options
+    song_list = get_data('perform').perform_options
+    @mindstate_goal = args.mindstate_goal.to_i
     skip_clean = args.noclean
     check_song_only = args.checksong
 
-    instrument = @settings.instrument
-    is_instrument_worn = instrument.nil?
+    worn_instrument = @settings.worn_instrument
```
Comment (rpherbig) [inline performance.lic:28]: I'm confused by the number of vars we're using here. Is my understanding right?
* `worn_instrument` - string, the instrument name which can be played while worn, optional
* `@instrument` - string, the instrument name which must be gotten before playing, optional
* `is_instrument_worn` - bool, whether or not the instrument can be played while worn, required
* `@instrument_out` - bool, whether or not we got an instrument, required
* `@hand_armor` - string, the hand armor to be removed if the instrument is _not_ worn, optional
* `@hand_armor_removed` - bool, whether or not the hand armor was removed, required

File: performance.lic [hunk 5]
```diff
@@ -52,37 +63,64 @@ class Performance
...
     Flags['ct-song'] = true
   end
 end
 
 before_dying do
   fput('stop play')
-
   instrument = get_settings.instrument
-  DRC.bput("stow #{instrument}", 'You put') if instrument
+  DRCI.stow_item?(instrument) if instrument && DRCI.in_hands?(instrument)
```
Comment (rpherbig) [inline performance.lic:123]: Is this still necessary now that we're cleaning up in `wear_hand_armor_and_put_away_instrument`?

File: performance.lic [hunk 6]
```diff
@@ -52,37 +63,64 @@ class Performance
...
+        if done_training?
+          wear_hand_armor_and_put_away_instrument
+          exit
+        end
       end
     end
   end
 
+  def remove_hand_armor_and_get_instrument
+    if DRCI.remove_item?(@hand_armor)
```
Comment (rpherbig) [inline performance.lic:87]: If either of these steps fail, should we exit the script?

File: performance.lic (hunk 1 above)
Comment (MahtraDR) [reply performance.lic:24]: Would the command line arg take precedence over the yaml setting in that OR? My understanding is that it wouldn't then be clear which one we're using.

Comment (MahtraDR) [reply performance.lic:24]: Nevermind, @asechrest explained this to me.

File: performance.lic (hunk 5 above)
Comment (MahtraDR) [reply performance.lic:122]: For the case when you kill performance manually, this makes sense, no?

Comment (rpherbig) [reply performance.lic:123]: I'm not sure. I mean, most scripts aren't designed to be killed at any arbitrary point in their execution (or at least, it's assumed they'll leave things in an unknown state). Most scripts don't try to clean up anything beyond Flags if they're killed. 

Should we re-wear our hand armor too? We don't have access to the instance var (I think), so it'd be a matter of either wearing it no matter what or checking if we're already wearing it and then trying to wear it?

File: performance.lic [hunk 7]
```diff
@@ -52,37 +63,64 @@ class Performance
...
           exit
         else
-          play_song?(@settings, @song_list, is_instrument_worn, skip_clean)
+          DRC.play_song?(@settings, song_list, is_instrument_worn, skip_clean)
         end
       when /Perhaps you should find somewhere drier/
-        exit
+          wear_hand_armor_and_put_away_instrument
+          exit
+        end
```
Comment (rpherbig) [inline performance.lic:80]: Is this an extra `end`?

**Pull request merged**